ES|QL: Improve aggregation over constants handling

astefan opened the pull request:
This change consists of:
- Add separate rule for dealing with nulls in aggregations
- Duplicate `SubstituteSurrogate` in "Operator Optimization" batch
- Many more tests
- Add test for `median_absolute_deviation` function
- Add mv handling to `top` function
- Allows `PropagateEvalFoldables` rule to also deal with aggregate functions

Addresses part of https://github.com/elastic/elasticsearch/issues/100634. Missing bits:
- `AwaitsFix`es from `LogicalPlanOptimizerTests` cannot be removed yet. This likely has also to do with `substitutions()` batch -> `NormalizeAggregate()` rule (that needs to be added?) in `LogicalPlanOptimizer`
- `median_absolute_deviation` is [still pending on having its own `mv_` function](https://github.com/elastic/elasticsearch/pull/112055)
- st_centroid agg `mv_*` function [still needs addressing](https://github.com/elastic/elasticsearch/issues/108215)
- recent update: [`mv_values` sister function is pending addition](https://github.com/elastic/elasticsearch/issues/112445)

Fixes https://github.com/elastic/elasticsearch/issues/110257
Fixes https://github.com/elastic/elasticsearch/issues/104430
Fixes https://github.com/elastic/elasticsearch/issues/100170

Needs more tests for the new rule and the existent ones in `LogicalPlanOptimizerTests`.

astefan committed:
Add separate rule for dealing with nulls in aggregations
Duplicate SubstituteSurrogate in "Operator Optimization" batch
Many more tests
Add tests for mad
Add mv handling to top function

astefan commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial.csv-spec:
```diff
@@ -37,6 +37,14 @@ wkt:keyword
+null             |null                     |null
+;
+
+########### failing :-( with InvalidArgumentException: Does not support yet aggregations over constants
```
This will be removed once we have mv_ function for st_centroid_agg.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Top.java:
```diff
@@ -197,11 +202,18 @@ public AggregatorFunctionSupplier supplier(List<Integer> inputChannels) {
     public Expression surrogate() {
```
The `surrogate` method, as it stands now, is more a "surrogate-expression-for-foldable-scenario" kind of method. This implies that the behavior that existed below before this change is not possible anymore.

costin replied:
Right - we should probably look into introducing a different interface altogether: surrogate was initially used for expressions that knew they'd be transformed.
But it evolved into a mechanism for "folding" however not to a value, but another expression (which itself might be foldable or not).

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -198,9 +199,11 @@ protected static Batch<LogicalPlan> operators() {
             new SplitInWithFoldableValue(),
             new PropagateEvalFoldables(),
             new ConstantFolding(),
+            new SubstituteSurrogates(),
```
`SubstituteSurrogates` is also executed in `Substitutions` batch from the LogicalPlanOptimizer, but in reality it shouldn't be part of that batch because it **relies on `ConstantFolding`, `FoldNull` and `PropagateEvalFoldables`** to be able to do its job properly.

This leads me to the same idea of `SurrogateExpression` being more an aid for foldables, rather than a surrogate expression for every kind of expression. The true (non-foldables) surrogate expression (like the one in `TOP` aggregate function) should only be used **after all the constants have been resolved**, propagated and replaced.

Removing `SubstituteSurrogates` from the `Substitutions` batch is currently not possible because `TranslateMetricsAggregate` needs it. This `metrics` specific rule is also the reason why `ReplaceStatsNestedExpressionWithEval` is executed twice in the `Substitutions` batch. One idea is to **investigate** and see if it's possible to change the implementation of `TranslateMetricsAggregate` so that is not forcing multiple uses of the same rule in that batch; maybe `TranslateMetricsAggregate` should be dealt with **like `inlinestats` deals with the split of plans**.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/ReplaceAggregatesWithNull.java:
```diff
@@ -0,0 +1,111 @@
+ * All aggregate functions that are also nullable (COUNT_DISTINCT and COUNT are exceptions), will get a NULL
+ * field replacement by the FoldNull rule, COUNT_DISTINCT will benefit from PropagateEvalFoldables.
+ */
+public final class ReplaceAggregatesWithNull extends OptimizerRules.OptimizerRule<Aggregate> {
```
This rule is a simplified variant of `SubstituteSurrogates`.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/SubstituteSurrogates.java:
```diff
@@ -44,21 +43,27 @@ public SubstituteSurrogates() {
+        Map<AggregateFunction, Attribute> aggFuncToAttr = new HashMap<>(); // existing aggregate and their respective attributes
+        List<Alias> transientEval = new ArrayList<>(); // surrogate functions eval
         boolean changed = false;
+        boolean hasSurrogates = false;
```
I've done this to shortcircuit the execution earlier in the execution.

astefan commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -5091,7 +5091,7 @@ public void testNullFoldingDoesNotApplyOnAggregate() throws Exception {
         assertEquals(countd, rule.rule(countd));
         countd = new CountDistinct(EMPTY, NULL, NULL);
-        assertEquals(new Literal(EMPTY, null, LONG), rule.rule(countd));
+        assertEquals(countd, rule.rule(countd));
```
This is the consequence of `CountDistinct` not being nullable anymore.

astefan committed:
One more test from one of the reported bugs

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/CountDistinct.java:
```diff
@@ -151,6 +152,11 @@ public DataType dataType() {
 
+    @Override
+    public Nullability nullable() {
+        return Nullability.FALSE;
```
👍

costin reviewed (approved):
LGTM - great tests and comments!

ivancea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Values.java:
```diff
@@ -115,4 +116,9 @@ public AggregatorFunctionSupplier supplier(List<Integer> inputChannels) {
+
+    @Override
+    public Expression surrogate() {
+        return field().foldable() ? field() : null;
```
Values not only merges values, but also removes duplicates (If no test was triggered because of this, we should add some!)

```
ROW x = [1, 1, 2] | STATS a = VALUES(x)

-> [1, 2]
```

astefan replied:
Good catch, BUT I think we have a problem with the documentation. It's not mentioning this aspect. There were other misses in our functions docs (which are fixed in this PR), I think we need to review our documentation on functions and double check its correctness and completeness. I will create an issue.

astefan replied:
@ivancea thank you for pro-actively checking this PR 🙏, that was very helpful.
I've created two issues:
- improving our documentation: https://github.com/elastic/elasticsearch/issues/112437
- this PR also now depends on pending `mv_values` addition: https://github.com/elastic/elasticsearch/issues/112445

alex-spies reviewed (commented):
This is great @astefan ! I think this change is sound and added mostly minor remarks.

My only major remark is: I think we need LogicalPlanOptimizerTests cases that prove that the foldable propagation actually takes place. The csv tests are _great_, but they do not prove that foldable propagation actually takes place, only that the result is correct.

But you already mentioned more optimizer tests as one of the tasks to un-draft :)

astefan committed:
Addressing reviews

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/ReplaceAggregatesWithNull.java:
```diff
@@ -0,0 +1,111 @@
+                && af.field().foldable()
+                && (DataType.isNull(af.field().dataType())
+                    || af.field() == null
+                    || af.field() instanceof Literal lit && lit.value() == null)) {
```
Yeah, you are right. `isNull` should do it.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/ReplaceAggregatesWithNull.java:
```diff
@@ -0,0 +1,111 @@
+                    || af.field() == null
+                    || af.field() instanceof Literal lit && lit.value() == null)) {
+                isNewAgg = true;
+                if (af instanceof CountDistinct || af instanceof Count) {
```
Right now, I don't have an opinion on this.
I tend to do centralized things in optimization rules where the behavior is more "controllable" so to speak. Having the null foldability behavior spread throughout aggregate functions assumes that aggregate functions authors are aware of the null foldability special handling. Pros and cons. I'll think about this some more when I'll look at refactoring this for a more [foldable-centric approach (vs. surrogate expressions approach)](https://github.com/elastic/elasticsearch/pull/112392/files#r1738501593).

astefan committed:
Make count_distinct deal with Validateable interface

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into aggregations_over_constants

astefan committed:
More count_distinct fixes, more tests for percentile's foldable second
argument

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into aggregations_over_constants

astefan committed:
Add more tests
Introduce isConstantFoldable() for aggregate functions

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into aggregations_over_constants

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

astefan committed:
Update docs/changelog/112392.yaml

elasticsearchmachine commented:
Hi @astefan, I've created a changelog YAML for you.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/AggregateFunction.java:
```diff
@@ -81,6 +81,10 @@ public List<? extends Expression> parameters() {
         return parameters;
     }
 
+    public boolean isConstantFoldable() {
```
This is a recent update. For now, I've introduced this one because of the geo aggregations that do not have, from what I've seen, a mechanism for dealing with constants.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -165,6 +170,7 @@ protected static Batch<LogicalPlan> operators() {
             new CombineBinaryComparisons(),
             new CombineDisjunctions(),
             new SimplifyComparisonsArithmetics(DataType::areCompatible),
+            // new ReplaceStatsAggExpressionWithEval(),
```
This is not a leftover. This is a test I did for https://github.com/elastic/elasticsearch/issues/113393. The rules in these two batches need to be cleaned up, unfortunately [not easy to do](https://github.com/elastic/elasticsearch/pull/112392/files#r1738501593).

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PropagateEvalFoldables.java:
```diff
@@ -48,10 +54,20 @@ public LogicalPlan apply(LogicalPlan plan) {
+            } else if (p instanceof Aggregate agg) {
+                List<NamedExpression> newAggs = new ArrayList<>(agg.aggregates().size());
+                agg.aggregates().forEach(e -> {
+                    if (Alias.unwrap(e) instanceof AggregateFunction af && af.isConstantFoldable()) {
```
`isConstantFoldable()` to skip geo aggregate functions. Otherwise, AbstractPhysicalOperationProviders will complain that  the geo functions having a foldable value as arguments reached [a point they shouldn't have](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java#L248).

astefan commented:
@costin, @alex-spies, @ivancea thank you for the preliminary check. I've fixed few more issues in the meantime, added more tests, addressed most of the comments and added LogicalPlanOptimizer tests.

I am not 100% confident that I fixed everything possible, but those scenarios that I could think of.
The PR has a long history and is rabbit-holey enough and there are several (easier and more difficult) follow-ups in the same ballpark unfortunately.

@costin also encountered the same issues as me with his [per agg filtering PR](https://github.com/elastic/elasticsearch/pull/113735) and the aggregate functions need a refactoring and change of approach regarding foldability and surrogate expressions. Until that happens (soon because things are getting too messy the more we work in the area of aggregations) my hope with this PR is that its tests will be the most important asset, then `PropagateEvalFoldables` gets to do its full job, third - the many bugs I fixed along the way in various places (`CountDistinct`, `Percentile`, surrogate for `Top` and `Values` etc).

astefan committed:
Address feedback
Add capability

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceAggregatesWithNull.java:
```diff
@@ -0,0 +1,106 @@
+                if (af instanceof CountDistinct || af instanceof Count) {
+                    value = 0L;
+                }
+            } else if (Expressions.isNull(e)) {// FoldNull can replace an entire aggregate function with a null Literal
```
> Does it make sense for FoldNull to still work on Aggregates? Maybe it's not worth changing, but as FoldNull would generate a wrong plan without this new rule, it would:
> 
> Make FoldNull not generate wrong plans
> Make this rule simpler (Moving the conditional to FoldNull really)

`FoldNull` has the job of quickly transforming expressions into `null` based on the **`Nullability` characteristic** of that expression. This is this rule **sole job**, it's contained, simple and well defined.

`Nullability` is a "shortcut" to quickly evaluate expressions to `null`. In the specific case of aggregation functions there are two exceptions to this `Nullability` thing - `count` and `count_distinct` - which can never be `null` (and I discovered the bug where we didn't mark `CountDistinct` as such, now it's fixed). If I were to change `FoldNull`, I would complicate its code with many exceptions to a general, simple rule. Maintaining this rule from this point on will be more complicated.

The alternative is to have this "aggregations on nulls scenarios" in a self contained rule - `ReplaceAggregatesWithNull` - which:
- contains the code of dealing with `null`s in aggs in a single place. If `FoldNull` would change, then this code would be split in two places - harder to maintain.
- deals with the "effect" of **potentially** having `FoldNull` run before it and inserting a `null` there

Your suggestion about `FoldNull` not doing the replacement for aggs is a valid one and will work if implemented, but my main argument in favor of the solution in this PR is that of **self-containment**, **clear separation of rules' tasks** and **ease of maintenance**.

> A small step towards having rules not depend on each other (Or at least not generate invalid plans)

Regarding this dependency, yes, it exists. And the dependency between optimization rules is a reality that spawns from the same arguments I mentioned above -  **self-containment**, **clear separation of rules' tasks** and **ease of maintenance**.

Each rule has one job and has some conditions that need to be `true` for the rule to be active. These conditions can be "activated" by the effect of other rules' executions.
If all the rules in the optimizer **would be independent**, each rule would perform the tasks that now can be found in **other rules** and the **code complexity and duplication** would be a nightmare.

I dare to say that **rule dependencies being inexistent is a fantasy**, we cannot have an easy to maintain code if this would be a reality.

ivancea replied:
> I dare to say that rule dependencies being inexistent is a fantasy, we cannot have an easy to maintain code if this would be a reality.

Yeah, I'm more into making a "best effort" in making that happen. Not that it should be that way everywhere. That is, having rules without dependencies as a default, and then documenting dependencies for those that need them.

Btw, I'm fine with the solution, thanks for explaining!

astefan replied:
> Yeah, I'm more into making a "best effort" in making that happen. Not that it should be that way everywhere. That is, having rules without dependencies as a default, and then documenting dependencies for those that need them.

Definitely. And this is actually the hardest part when it comes to these rules: finding the best place to make the change so that rules remain self-contained, their job is as well defined as it can be and the change just "fits" in the ecosystem of the other rules. Of course, if another rule does something it shouldn't, that rule should change and maybe move what it was doing to somewhere else.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceAggregatesWithNull.java:
```diff
@@ -0,0 +1,106 @@
+        for (NamedExpression agg : aggs) {
+            Expression e = Alias.unwrap(agg);
+            Object value = null;
+            boolean isNewAgg = false;
```
You're right. The names are not good. I'll change them.

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into aggregations_over_constants

astefan committed:
Merge branch 'aggregations_over_constants' of https://github.com/astefan/elasticsearch into aggregations_over_constants

astefan committed:
More capabilities to csv-spec files

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into aggregations_over_constants

astefan committed:
More bwc checks

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into aggregations_over_constants

ivancea reviewed (approved):
LGTM!

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:
I can see that `from employees | eval x = [1,2,3], y = 5 + 6 | stats by yy = y + 1` doesn't even work on main, so that's a good thing to fix in a later PR.

alex-spies replied:
Opened https://github.com/elastic/elasticsearch/issues/114714

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats_percentile.csv-spec:
```diff
@@ -1,12 +1,74 @@
+       p50 = 40+10 
+| stats percentile(value200, p90), 
+        percentile(value200, p100), 
+        percentile(value200, p3_null), 
```
Shouldn't `percentile(x, null)` be invalid? Similarly for `count_distinct(x, null)`.

astefan replied:
That's a good question. Here's the paradox: `percentile` value (the second argument) must be a number between 0 and 100. If it's 1000 we are issuing a warning and treat the result of the agg as `null`; by the principle that we issue warnings and return `null`s when things "don't work" I think we should treat `percentile` as `null` if the second parameter is `null`.

Yes, it's not a value between 0 and 100, it's not a missing value, it's.... "unknown" value, a `null`. Which is a bit special than missing or a number between 0 and 100. In `main` this query fails with an ugly exception.

There are other problems or, better said, inconsistencies: `weighted_avg` specifically forbids `null` at data type validation time (`weighted_avg(x, null)`), but it doesn't do it when the value received is the result of an evaluation (`eval n = null + 2 | stats weighted_avg(x, n)`. If `weighted_avg` does check for `null` and forbids it explicitly, `percentile` doesn't even do it.

astefan replied:
I don't have an answer for you to this question right now. There should be a generic approach to dealing with aggregations' arguments that accept only constants: `top`, `percentile`, `weighted_avg`.

costin replied:
The behavior is inconsistent and needs to be fixed, though not in this PR.
If the arguments have to be literals, then throwing an exception (invalid query) would be the way to go.
If not, then folding+ warning is the way to go.

For the former, the function implementation would have to be improved to rely on Validatable interface. Another way, would be marking the field as foldable/not-null and doing this in the planner as it will work across all functions.

alex-spies replied:
We don't have to hash this out now, but _maybe_ it'd be safer to start with a validation exception now - we can still go back and return null later, while the other way around _could_ be considered a breaking change, albeit in a very edge case scenario.

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats_count_distinct.csv-spec:
```diff
@@ -177,3 +177,70 @@ m:long  | languages:i
 10   | null
 ;
+
+foldablePrecision
+required_capability: extend_aggs_on_constants_support
+from employees | eval x = 82+8 | stats count_distinct(salary, x*100);
+
+count_distinct(salary, x*100):long
+100
```
I think the data set is too small to test that the precision is actually correctly propagated. Shouldn't we add a test on a dataset where different values for the precision lead to different `count_distinct` results?

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats_top.csv-spec:
How about some tests where the limit and/or the `"asc"/"desc"` is propagated? And, similarly, tests for `weighted_avg` where the weights are propagated?

astefan replied:
I've tried that but the situation is problematic because of the "surrogate" expressions abuse. Right now the validation of these constant arguments in `percentile` and `weighted_avg` is done in `resolveType` which is not completely correct. It should be done in `validate` after any constant values are propagated, folded and replaced.

**If** we do this in `validate` then in some cases the validation is not happening at all because `weighted_avg` **disappears**, being replaced by one of the formulas in `surrogate`. For example `ROW values=[1,2,3,4,5] | weighted_avg(values, 0)` - if we do the validation properly in `validate` that method won't even be called because when it is there is no `WeightedAvg` instance anymore (it's getting replaced with `MvAvg(s, field)`). Unless we do something about aggregation foldability and surrogate expressions we are in a deadlock: we fix something but break something else.

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats_percentile.csv-spec:
```diff
@@ -1,3 +1,22 @@
+;
+
+percentileOfNullsOnRealIndex
+from employees | eval x = null::integer, y = null, z = 123 - null | stats percIntNull = percentile(y, 90), percNull = percentile(y, 90), percentile(null, 90), percentile(null+2, 90), percentile(z, 90) by languages | sort languages desc;
```
I think the test fix got lost :) It's only reformatted in the current version, but the definitions of `percIntNull` and `percNull` are still the same.

alex-spies reviewed (commented):
I have a small round of comments - plan to finalize the re-review tomorrow :)

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
`weighted_avg` and `top` are special in that they can take 2(+) arguments which, in case of `weighted_avg`,  _can_ even be non-foldable. I think we should add tests that show foldable propagation and propagation of nulls into `weighted_avg` and `top` as well.

alex-spies reviewed (changes requested):
I think this is a very good PR that we should go forward with, after resolving whatever conflicts may arise from the `STATS ... WHERE ...` support that was merged today.

The only thing I think this is really missing is optimizer tests that demonstrate that foldable propagation (of non-null expressions) actually takes place. And we should double check if we really want `percentile(x, null)` and `count_distinct(x, null)` to just be `null` instead of an invalid query, as that's a decision we won't be able to take back without becoming a breaking change.

Other than that, my remarks are mostly minor.

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial.csv-spec:
```diff
@@ -37,6 +37,18 @@ wkt:keyword
+FROM airports 
+| eval z = TO_GEOPOINT(null) 
+| STATS centroidNull = ST_CENTROID_AGG(null), 
+        centroidExpNull = ST_CENTROID_AGG(TO_GEOPOINT(null::string)), 
```
Per https://github.com/elastic/elasticsearch/issues/106025, returning `null` is correct!

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into aggregations_over_constants

astefan committed:
Cleanup after merging "main"

astefan committed:
Skip one more test

astefan committed:
Change one test according to reviews

astefan commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats_count_distinct.csv-spec:
```diff
@@ -177,3 +177,70 @@ m:long  | languages:i
+from employees | eval x = 82+null | stats count_distinct(salary, x*100), count_distinct(salary, null), count_distinct(salary, null + 1);
+
+count_distinct(salary, x*100):long|count_distinct(salary, null):long|count_distinct(salary, null + 1):long
+100                               |100                              |100
```
I am not so sure :-), now that you mentioned this aspect. Maybe we want null here, but at the same time `count_distinct` and `count` DO make sense to be not nullable, counting something should always be something concrete.

If `null` is not the default precision and we cannot return `null` as the return value of `count_distinct`, should we error out then? (like weighted_avg and percentile)

costin replied:
count are _special_ in that they are not nullable - they always return a value. If anything, MV_COUNT should be aligned with it.
It's up to the function to decide if null is treated as "default value" or invalid. Since there's no COUNT_DISTINCT(field), I'd opt for null meaning default.

alex-spies replied:
I'd be in favor of `COUNT_DISTINCT(field, null)` resulting in an invalid query exception via the `Validatable` interface - but I'm fine with default precision, as long as we document this clearly :)

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceAggregatesWithNull.java:
```diff
@@ -46,48 +45,44 @@ public ReplaceAggregatesWithNull() {
+            if (remainingAggregates.isEmpty() == false) {// build the new Aggregate with the rest (non-null) aggregates
+                plan = new Aggregate(source, aggregate.child(), aggregate.aggregateType(), aggregate.groupings(), remainingAggregates);
             } else {
                 // All aggs actually have been optimized away
```
I am not following and I need more details. `remainingAggregates` being empty means exactly that - the `BY` part is empty, it was folded no `null` by this rule. If `BY` is reduced to `null` this means only one row in the results.

alex-spies replied:
Actually, I found a reproducer on your branch:
```
ROW field = null::integer, otherfield = [1,2] | STATS min(field) by otherfield | DROP otherfield

  min(field)   
---------------
null 
```
This should instead be
```
  min(field)   
---------------
null 
null
```
(as it is on main) because there 2 values for `otherfield` and thus 2 groups, resp. 2 output rows from the `STATS`.

It's insufficient to look at the aggregates. We need to check `aggregate.groupings()` instead, as dropping the groupings after the stats lets us optimize them away from the aggregates.

The removal of the groupings from the aggregates happens in `CombineProjections`, like in the example above:
```
[2024-10-17T11:25:35,813][INFO ][o.e.x.e.o.LogicalPlanOptimizer] [runTask-0] Rule logical.CombineProjections applied
Limit[1000[INTEGER]]                                                                                     = Limit[1000[INTEGER]]
\_EsqlProject[[min(field){r}#7]]                                                                         ! \_Aggregate[STANDARD,[otherfield{r}#4],[MIN(field{r}#2,true[BOOLEAN]) AS min(field)]]
  \_Aggregate[STANDARD,[otherfield{r}#4],[MIN(field{r}#2,true[BOOLEAN]) AS min(field), otherfield{r}#4]] !   \_Row[[TOINTEGER(null[NULL]) AS field, [1, 2][INTEGER] AS otherfield]]
    \_Row[[TOINTEGER(null[NULL]) AS field, [1, 2][INTEGER] AS otherfield]]                               ! 
```

astefan replied:
I don't think `CombineProjections` is to blame, it is doing the right thing. This can be seen with queries that have their projections correctly pruned away.

The "workaround" of calling `SubstitueSurrogates` again in the "Operator Optimization" is to blame here and this shows the vicious cycle of fixing something - breaking something else that [I mentioned before](https://github.com/elastic/elasticsearch/pull/112392/files#r1802734728). Foldability and surrogate expressions represent a faulty principle and when things start to not make sense and the code "struggles" to make the right thing (by using workarounds like calling `SubstituteSurrogates` twice) then it is clear that something more fundamental is not right.

At this point there are two options available:
- create an issue for this and merge the PR for the benefits of the tests it adds. This though introduces a bug in a edgy situation (the one that we know of, there may be others that we don't)
- try to make `SubstituteSurrogates` don't do its thing for aggregations that have a grouping which is not kept as a projection anymore. This is clearly a workaround that tries to ignore what `CombineProjections` did. I am trying now this second idea.

astefan committed:
More reviews

astefan commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -5565,7 +5565,353 @@ public void testToDatePeriodToTimeDurationWithField() {
 
-    private Literal nullOf(DataType dataType) {
-        return new Literal(Source.EMPTY, null, dataType);
+    public void testCheckPercentileFoldableSecondArgument() {
```
Agree. It must be a ValidatableTests kind of class, because these tests are specific to this approach of performing the validation. I've moved them to a new class. We need a followup to clean the rest of the "Verification" exception checks.

astefan committed:
Bug fix related to the source() used when creating the Validations
error message

alex-spies reviewed (commented):
Heya, just wanted to pick this up again and summarize what I think we need to do:
- Per the discussion with Costin, `percentile(field, null)` is still not well defined (null vs invalid query) - okay to hash this out in a follow-up but maybe invalidating for now is safer w.r.t. bwc.
- `ST_CENTROID_AGG(null)` should return `null` instead of `Point(NaN NaN)`.
- Some additional test cases won't hurt.
- Fixing [this edge case](https://github.com/elastic/elasticsearch/pull/112392#discussion_r1804426669) where all agg functions are optimized away

Consider this unblocked from my side as my main reason for requesting changes was the discussion about `percentile(field, null)` and similar cases. We could solve some problems in follow-up PRs as well, as I think the general approach here works :)

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into aggregations_over_constants

astefan committed:
Fix some things after pull from main

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into aggregations_over_constants

felixbarny commented:
Is this PR superseded by https://github.com/elastic/elasticsearch/pull/139797 or is it something we're planning to do additionally?

wchaparro commented:
<!-- analytics-pr-review-management:stall-warn -->

Hi @astefan — this PR has been identified as **stalled** (open and without meaningful updates for 30+ days).

Please address the requested changes when you can.

It will be **closed in two weeks** if there is no further activity. Please reply, push an update, or close it yourself if it is no longer needed.

— Analytics PR review management
